pkg/patterns/declarative/pkg/applier/direct_test.go

```diff
@@ -0,0 +1,204 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package applier
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/api/meta/testrestmapper"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/rest/fake"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	kubectltesting "k8s.io/kubectl/pkg/cmd/testing"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+type directApplierTestSite struct {
+	Error     error
+	applyOpts *apply.ApplyOptions
+}
+
+func (d *directApplierTestSite) Run(a *apply.ApplyOptions) error {
+	d.applyOpts = a
+	return nil
+}
+
+// same manner with a function newUnstructuredDefaultBuilderWith on k8s.io/cli-runtime/pkg/resource/builder_test.go
+func (d *directApplierTestSite) NewBuilder(opt ApplierOptions) *resource.Builder {
+	return resource.NewFakeBuilder(
+		func(version schema.GroupVersion) (resource.RESTClient, error) {
+			return &fake.RESTClient{}, nil
+		},
+		func() (meta.RESTMapper, error) {
+			return testrestmapper.TestOnlyStaticRESTMapper(scheme.Scheme), nil
+		},
+		func() (restmapper.CategoryExpander, error) {
+			return resource.FakeCategoryExpander, nil
+		})
+}
+
+func (d *directApplierTestSite) NewFactory() cmdutil.Factory {
+	return kubectltesting.NewTestFactory()
+}
+
+func newDirectApplierTest(d *directApplierTestSite) Applier {
+	return &DirectApplier{inner: d}
+}
+
+func TestDirectApply(t *testing.T) {
+	tests := []struct {
+		name               string
+		namespace          string
+		manifest           string
+		validate           bool
+		args               []string
+		err                error
+		expectApplyOptions *apply.ApplyOptions
+		expectCheckFunc    func(opt *apply.ApplyOptions) error
+	}{
+		{
+			name:      "manifest",
+			namespace: "",
+			manifest: `---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-operator
+  namespace: kube-system`,
+			expectApplyOptions: &apply.ApplyOptions{},
+			expectCheckFunc: func(opt *apply.ApplyOptions) error {
+				return nil
+			},
+		},
+		{
+			name:      "manifest with namespace",
+			namespace: "test-namespace",
+			manifest: `---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-operator
+  namespace: kube-system`,
+			expectApplyOptions: &apply.ApplyOptions{
+				Namespace: "test-namespace",
+			},
+			expectCheckFunc: func(opt *apply.ApplyOptions) error {
+				if opt.Namespace == "test-namespace" {
+					return nil
+				} else {
+					return fmt.Errorf("namespace doesn't match to \"test-namespace\"")
+				}
+			},
+		},
+		// This test use fake test factory on "k8s.io/kubectl/pkg/cmd/testing", that's why this test for validate is commented out.
+		// opt.Validator is always set to validation.NullSchema instance by fake test factory, so we can't check if validator is set successfully.
```

Comment:
We should this test w/o live k8s cluster, so I use fake factory.
Unit test become to pass by using that, but we can't check validate options.. :thinking: 
Is there any good implementations?

Reply:
Sorry for delay!  I just uploaded a test with a fake kubeapiserver that is being developed across a few projects using copy-and-paste reuse while we figure it out!  (The projects being kOps, kpt and this one!)

I still want this PR as I agree we need more test coverage, but I think knowing that the other PR can cover the HTTP round trips makes not validating here OK.

Although actually, validation is generally moving to the server, so we might be OK to just discourage client-side validation anyway...